Support grouping results by severity

cx-james-bostock opened the pull request:
By default, results are grouped by vulnerability and filename. This change allows the grouping to optionally also include the result
severity (i.e., two results for the same vulnerability in the same file but with different severities will be considered as two distinct results).

This is a reimplementation of a change made for Woolworths. The reimplementation makes the change in functionality configurable via the group-by-severity property in the application.yml file. If this property is not specified, the normal behavior is exhibited (grouping only by vulnerability and filename).

If this PR is accepted, I will create a separate one to update the CxFlow documentation accordingly.

cx-james-bostock committed:
Support grouping results by severity

By default, results are grouped by vulnerability and filename. This
change allows the grouping to optionally also include the result
severity (i.e., two results for the same vulnerability in the same
file but with different severities will be considered as two distinct
results).

cx-alexey-kononov commented on src/main/java/com/checkmarx/sdk/config/CxProperties.java:
```diff
@@ -52,6 +52,8 @@
 
     private Map<String, String> customStateMap;
 
+    private Boolean groupBySeverity = false;
```
Can you please add a comment to this field?

cx-james-bostock replied:
Will do.

cx-alexey-kononov commented on src/main/java/com/checkmarx/sdk/service/CxService.java:
```diff
@@ -755,6 +755,8 @@ private boolean filterOsa(List<Filter> filters, CxOsa osa) {
                         if(!resultType.getFalsePositive().equalsIgnoreCase("FALSE")){
                             falsePositive = true;
                         }
+                        log.debug("getIssues: id: {}, severity: {}, result type filename: {}, result type severity: {}",
```
This message will add quite a lot of logging. Are you sure you want to have it here?
I remember there is a similar per-issue log somewhere in CxFlow ResultsService during publishing.

cx-james-bostock replied:
This was left over from when I was testing the change. I will remove it.

cx-alexey-kononov commented on src/main/java/com/checkmarx/sdk/dto/ScanResults.java:
```diff
@@ -256,7 +258,12 @@ public boolean equals(Object o) {
-            return filename.equals(issue.filename);
+            if (groupBySeverity) {
+        	if (!filename.equals(issue.filename)) return false;
+        	return severity.equals(issue.severity);
```
```suggestion
        	return filename.equals(issue.filename) && severity.equals(issue.severity);
```

cx-james-bostock replied:
Yes, this looks better. Will make this change.

cx-alexey-kononov commented on src/main/java/com/checkmarx/sdk/dto/ScanResults.java:
```diff
@@ -256,7 +258,12 @@ public boolean equals(Object o) {
 
             if (!vulnerability.equals(issue.vulnerability)) return false;
-            return filename.equals(issue.filename);
+            if (groupBySeverity) {
```
Please auto-format.

cx-alexey-kononov commented on src/main/java/com/checkmarx/sdk/config/CxProperties.java:
```diff
@@ -259,5 +261,13 @@ public String getStateFullName(String key){
 	return stateFullName;
     }
+
+    public Boolean getGroupBySeverity() {
+	return groupBySeverity;
+    }
+
+    public void setGroupBySeverity(Boolean groupBySeverity) {
+	this.groupBySeverity = groupBySeverity;
+    }
```
Please auto-format.

cx-james-bostock replied:
Ditto.

cx-alexey-kononov reviewed (commented):
The new grouping logic is quite fundamental to CxFlow and should be tested thoroughly.

Specifically, we should check whether groupBySeverity works well with things like filters, thresholds (issue counting) or pull request formatting (table rows).

cx-james-bostock committed:
Updates in response to pull request review comments

Add comment for the CxPropeties.groupBySeverity member variable.

Remove debug log message.

Minor changes and whitespace cleanup to the ScanResults.equals()
method.

cx-james-bostock commented:
Hi @alex-ko-dev 

Regarding testing, at present, for a single use case, I am testing that the OOTB CxFlow behavior is preserved and that the new functionality matches that of the existing hard-coded customization.

The customer for whom this development was done is currently using it in production (i.e. with a custom build of CxFlow).

That said, it would be good to have automated tests to cover this and I am happy to try to add some. That said, if I try to run the SDK unit tests, none are run:

```
$ ./mvnw test
...
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0
```

cx-alexey-kononov commented:
Hi @james-bostock-cx , not sure why CxSDK tests don't run... Needs to be checked.
In any case, we have much better test coverage in CxFlow rather than SDK. So it's better to add/modify tests for the new functionality right in CxFlow.

cx-james-bostock commented:
Hi Alexey,

Many thanks!

Do the SDK tests run for you?

Regards,

James Bostock | Professional Services Engineer, APAC | ***@***.******@***.***> | +61 429 546 058 | Checkmarx<https://www.checkmarx.com/>

From: AlexeyK ***@***.***>
Sent: Wednesday, 25 August 2021 5:22 PM
To: checkmarx-ltd/checkmarx-spring-boot-java-sdk ***@***.***>
Cc: James Bostock ***@***.***>; Mention ***@***.***>
Subject: Re: [checkmarx-ltd/checkmarx-spring-boot-java-sdk] Support grouping results by severity (#153)


Merged #153<https://github.com/checkmarx-ltd/checkmarx-spring-boot-java-sdk/pull/153> into develop.

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/checkmarx-ltd/checkmarx-spring-boot-java-sdk/pull/153#event-5204824995>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASPCDQVFDDDKSUQSPYOTRWLT6SKZBANCNFSM5BZSVJUQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.
